tests/dualtor_io/test_link_failure.py

```diff
@@ -131,3 +132,122 @@ def test_standby_link_down_downstream_standby(
         expected_standby_host=lower_tor_host,
         expected_standby_health='unhealthy'
     )
+
+
+def test_active_tor_downlink_down_upstream(
+    upper_tor_host, lower_tor_host, send_server_to_t1_with_action,
+    toggle_all_simulator_ports_to_upper_tor,
+    shutdown_upper_tor_downlink_intfs
+):
+    """
+    Send traffic from server to T1 and shutdown the active ToR downlink on DUT.
+    Verify switchover and disruption lasts < 1 second
+    """
+    send_server_to_t1_with_action(
+        upper_tor_host, verify=True, delay=MUX_SIM_ALLOWED_DISRUPTION_SEC,
+        allowed_disruption=3, action=shutdown_upper_tor_downlink_intfs
+    )
+    verify_tor_states(
+        expected_active_host=lower_tor_host,
+        expected_standby_host=upper_tor_host,
+        expected_standby_health='unhealthy'
+    )
+
+def test_active_tor_downlink_down_downstream_active(
+    upper_tor_host, lower_tor_host, send_t1_to_server_with_action,
+    toggle_all_simulator_ports_to_upper_tor,
+    shutdown_upper_tor_downlink_intfs
+):
+    """
+    Send traffic from T1 to active ToR and shutdown the active ToR downlink on DUT.
+    Verify switchover and disruption lasts < 1 second
+    """
+    send_t1_to_server_with_action(
+        upper_tor_host, verify=True, delay=MUX_SIM_ALLOWED_DISRUPTION_SEC,
+        allowed_disruption=3, action=shutdown_upper_tor_downlink_intfs
+    )
+    verify_tor_states(
+        expected_active_host=lower_tor_host,
+        expected_standby_host=upper_tor_host,
+        expected_standby_health='unhealthy'
+    )
+
+
+def test_active_tor_downlink_down_downstream_standby(
+    upper_tor_host, lower_tor_host, send_t1_to_server_with_action,
+    toggle_all_simulator_ports_to_upper_tor,
+    shutdown_upper_tor_downlink_intfs
+):
+    """
+    Send traffic from T1 to standby ToR and shutdown the active ToR downlink on DUT.
+    Verify switchover and disruption lasts < 1 second
+    """
+    send_t1_to_server_with_action(
+        lower_tor_host, verify=True, delay=MUX_SIM_ALLOWED_DISRUPTION_SEC,
+        allowed_disruption=3, action=shutdown_upper_tor_downlink_intfs
+    )
+    verify_tor_states(
+        expected_active_host=lower_tor_host,
+        expected_standby_host=upper_tor_host,
+        expected_standby_health='unhealthy'
+    )
+
+
+def test_standby_tor_downlink_down_upstream(
+    upper_tor_host, lower_tor_host, send_server_to_t1_with_action,
+    toggle_all_simulator_ports_to_upper_tor,
+    shutdown_lower_tor_downlink_intfs
+):
+    """
+    Send traffic from server to T1 and shutdown the standby ToR downlink on DUT.
+    Verify no switchover and no disruption
+    """
+    send_server_to_t1_with_action(
+        upper_tor_host, verify=True, delay=MUX_SIM_ALLOWED_DISRUPTION_SEC,
+        allowed_disruption=2, action=shutdown_lower_tor_downlink_intfs
```

Comment:
Why allowed_disruption is equal to 2 when expecting no disruption?

Reply:
> Why allowed_disruption is equal to 2 when expecting no disruption?

@theasianpianist Could you help with this question? I followed the allowed_disruption values from the existing cases defined by you. For upper tor it is 3 and for lower tor it is 2

Reply:
The reasoning for allowing multiple disruptions for some cases can be found here: https://github.com/Azure/sonic-mgmt/pull/4169

However, I'm not sure that same reasoning applies to these new cases since we are setting admin down on the ToR itself. If the link manager is able to see this admin down setting, it should prevent the described behavior. I'd suggest changing these values to 1 and seeing if the tests are still able to pass consistently.

Reply:
Set allow_disruption to 0 for 6 new test cases, all case passed.

```
=== Running tests in groups ===
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
============================================================================ test session starts =============================================================================
platform linux2 -- Python 2.7.17, pytest-4.6.5, py-1.11.0, pluggy-0.13.1
ansible: 2.8.12
rootdir: /data/sonic-mgmt-int/tests, inifile: pytest.ini
plugins: forked-1.3.0, metadata-1.11.0, xdist-1.28.0, html-1.22.1, repeat-0.9.1, allure-pytest-2.8.22, ansible-2.2.2
collecting ... ['conf-name', 'group-name', 'topo', 'ptf_image_name', 'ptf', 'ptf_ip', 'ptf_ipv6', 'server', 'vm_base', 'dut', 'inv_name', 'auto_recover', 'comment']
Finished testbed info generating.
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
collected 12 items                                                                                                                                                           

dualtor_io/test_link_failure.py::test_active_link_down_upstream PASSED                                                                                                 [  8%]
dualtor_io/test_link_failure.py::test_active_link_down_downstream_active PASSED                                                                                        [ 16%]
dualtor_io/test_link_failure.py::test_active_link_down_downstream_standby PASSED                                                                                       [ 25%]
dualtor_io/test_link_failure.py::test_standby_link_down_upstream PASSED                                                                                                [ 33%]
dualtor_io/test_link_failure.py::test_standby_link_down_downstream_active PASSED                                                                                       [ 41%]
dualtor_io/test_link_failure.py::test_standby_link_down_downstream_standby PASSED                                                                                      [ 50%]
dualtor_io/test_link_failure.py::test_active_tor_downlink_down_upstream PASSED                                                                                         [ 58%]
dualtor_io/test_link_failure.py::test_active_tor_downlink_down_downstream_active PASSED                                                                                [ 66%]
dualtor_io/test_link_failure.py::test_active_tor_downlink_down_downstream_standby PASSED                                                                               [ 75%]
dualtor_io/test_link_failure.py::test_standby_tor_downlink_down_upstream PASSED                                                                                        [ 83%]
dualtor_io/test_link_failure.py::test_standby_tor_downlink_down_downstream_active PASSED                                                                               [ 91%]
dualtor_io/test_link_failure.py::test_standby_tor_downlink_down_downstream_standby PASSED                                                                              [100%]

--------------------------------------------------------- generated xml file: /data/sonic-mgmt-int/tests/logs/tr.xml ---------------------------------------------------------
======================================================================== 12 passed in 4011.91 seconds ========================================================================
INFO:root:Can not get Allure report URL. Please check logs
```

Reply:
@theasianpianist Actually, for active tor, from test summary log, the real total_disruption is 1, for standby, it's 0.
Why all new 6 test cases are passed even setting `allowed_disruption = 0`? It's because if set `allowed_disruption = 0`, the following code will set it to 1 in line 233 of `data_plane_utils.py`. Is it fine?
```
        if delay and not allowed_disruption:
            allowed_disruption = 1
```

Reply:
@theasianpianist As we talked offline, I set allowed_disruption to 1 for Active Tor cases and set allowed_disruption to 0 for Standby Tor cases, after retesting, all test cases all passed.